Web/Controller/Companies.hs

```diff
@@ -0,0 +1,89 @@
+module Web.Controller.Companies where
+
+import Web.Controller.Prelude
+import Web.View.Companies.Index
+import Web.View.Companies.New
+import Web.View.Companies.Edit
+import Web.View.Companies.Show
+
+-- @todo: Add to Guide
+import qualified Network.Wai.Parse as Wai
+
+
+instance Controller CompaniesController where
+    action CompaniesAction = do
+        companies <- query @Company |> fetch
+        render IndexView { .. }
+
+    action NewCompanyAction = do
+        let company = newRecord
+        render NewView { .. }
+
+    action ShowCompanyAction { companyId } = do
+        companyWithRecords <- fetchCompanyWithRecords companyId
+
+        let uploadedFile = companyWithRecords.uploadedFile
+        uploadedFile <- getTemporaryDownloadUrlFromFile uploadedFile
+
+        -- @todo: Why is this not working?
+        -- companyWithRecords |> set #uploadedFile uploadedFile
+
+        render ShowView { companyWithRecords = companyWithRecords {uploadedFile = uploadedFile} }
+
+    action EditCompanyAction { companyId } = do
+        company <- fetch companyId
+        render EditView { .. }
+
+    action UpdateCompanyAction { companyId } = do
+        company <- fetch companyId
+        company
+            |> buildCompany
+            |> ifValid \case
+                Left company -> render EditView { .. }
+                Right company -> do
+                    company <- company |> updateRecord
+                    setSuccessMessage "Company updated"
+                    redirectTo EditCompanyAction { .. }
+
+    action CreateCompanyAction = do
+        uploadedFile <- withTransaction do
```

Comment:
What's the reason for `withTransaction` here?

Reply:
The transaction ensures that when the file upload files for any reason, there will be no leftover record in the `uploaded_files` table. In this case here, the transaction can likely be avoided by moving the `createRecord` to a later point (basically replacing the `updateRecord` call there)